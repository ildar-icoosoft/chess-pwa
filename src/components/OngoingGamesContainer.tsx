import React, { FC, useEffect, useState } from "react";
import { GamePreviewsList } from "./GamePreviewsList";
import Game from "../interfaces/Game";
import { getOngoingGames } from "../services/api";

export interface OngoingGamesContainerProps {
  onLoad?: () => void;
}

export const OngoingGamesContainer: FC<OngoingGamesContainerProps> = () => {
  const [games, setGames] = useState<Game[]>([]);

  useEffect(() => {
    getOngoingGames().then((res) => {
      setGames(res);
    });
  }, []);

  return <GamePreviewsList games={games} />;
};
