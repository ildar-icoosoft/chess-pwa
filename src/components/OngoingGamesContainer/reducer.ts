import { findIndex } from "lodash";
import Game from "../../interfaces/Game";

export interface State {
  games: Game[];
}

export type Action =
  | { type: "GET_GAMES"; payload: Game[] }
  | { type: "UPDATE_GAME"; payload: Partial<Game> }
  | { type: "CREATE_GAME"; payload: Game };

export const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "GET_GAMES":
      return {
        games: action.payload,
      };
    case "UPDATE_GAME": {
      const index: number = findIndex(state.games, {
        id: action.payload.id,
      });

      if (index !== -1) {
        return {
          games: state.games.map((item, itemIndex) => {
            if (itemIndex === index) {
              return {
                ...item,
                ...action.payload,
              };
            }
            return item;
          }),
        };
      }
      return state;
    }
    case "CREATE_GAME": {
      return {
        games: [...state.games, action.payload],
      };
    }
    default:
      throw new Error(`unknown action type ${(action as any).type}`);
  }
};
