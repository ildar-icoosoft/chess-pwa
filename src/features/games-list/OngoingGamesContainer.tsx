import React, { FC } from "react";
import { denormalize } from "normalizr";
import { useSelector } from "react-redux";
import { GamePreviewsList } from "./GamePreviewsList";
import { RootState } from "../../app/rootReducer";
import gameSchema from "../../normalizr/schemas/gameSchema";
import Game from "../../interfaces/Game";

const OngoingGamesContainer: FC<unknown> = () => {
  const games = useSelector((state: RootState) =>
    denormalize(
      Object.keys(state.entities.games),
      [gameSchema],
      state.entities
    ).filter((game: Game) => game.status === "started")
  );

  return <GamePreviewsList games={games} />;
};

export default OngoingGamesContainer;
