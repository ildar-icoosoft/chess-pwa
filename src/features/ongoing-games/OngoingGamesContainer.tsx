import React, { FC, useEffect } from "react";
import { denormalize } from "normalizr";
import { useDispatch, useSelector } from "react-redux";
import { GamePreviewsList } from "./GamePreviewsList";
import { RootState } from "../../app/rootReducer";
import { fetchOngoingGames } from "./ongoingGamesSlice";
import gameSchema from "../../normalizr/schemas/gameSchema";

const OngoingGamesContainer: FC<unknown> = () => {
  const dispatch = useDispatch();

  const items = useSelector((state: RootState) =>
    denormalize(state.ongoingGames.items, [gameSchema], state.entities)
  );

  useEffect(() => {
    dispatch(fetchOngoingGames());
  }, [dispatch]);

  return <GamePreviewsList games={items} />;
};

export default OngoingGamesContainer;
