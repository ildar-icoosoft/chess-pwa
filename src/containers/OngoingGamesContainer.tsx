import React, { FC, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { GamePreviewsList } from "../components/GamePreviewsList";
import { RootState } from "../app/rootReducer";
import { fetchGames, watchGames } from "../slices/gamesSlice";

const OngoingGamesContainer: FC<unknown> = () => {
  const dispatch = useDispatch();

  const { items, isLoading, error } = useSelector(
    (state: RootState) => state.games
  );

  useEffect(() => {
    dispatch(fetchGames());
    dispatch(watchGames());
  }, [dispatch]);

  return <GamePreviewsList games={items} />;
};

export default OngoingGamesContainer;
