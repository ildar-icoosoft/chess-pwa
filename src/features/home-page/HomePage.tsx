import React, { FC } from "react";
import { Button, Modal } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import OngoingGamesContainer from "../ongoing-games/OngoingGamesContainer";
import ChallengeAiFormContainer from "../challenge-ai-modal/ChallengeAiFormContainer";

import { RootState } from "../../app/rootReducer";
import {
  hideChallengeAiModal,
  showChallengeAiModal,
} from "../challenge-ai-modal/challengeAiModalSlice";
import { AppDispatch } from "../../app/store";

const HomePage: FC<unknown> = () => {
  const dispatch = useDispatch<AppDispatch>();

  const { isChallengeAiModalVisible } = useSelector(
    (state: RootState) => state.challengeAiModal
  );

  return (
    <>
      <Modal
        show={isChallengeAiModalVisible}
        onHide={() => dispatch(hideChallengeAiModal())}
        animation={false}
      >
        <Modal.Header closeButton>
          <Modal.Title>Play with the computer</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <ChallengeAiFormContainer />
        </Modal.Body>
      </Modal>
      <Button
        variant="primary"
        onClick={() => dispatch(showChallengeAiModal())}
      >
        Play with the computer
      </Button>
      <OngoingGamesContainer />
    </>
  );
};

export default HomePage;
