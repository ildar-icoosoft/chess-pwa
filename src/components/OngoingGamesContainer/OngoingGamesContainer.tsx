import React, { FC, useEffect, useReducer } from "react";
import { GamePreviewsList } from "../GamePreviewsList";
import { getOngoingGames, watchGames } from "../../services/api";
import { reducer } from "./reducer";

export type OngoingGamesContainerProps = Record<string, unknown>;

export const OngoingGamesContainer: FC<OngoingGamesContainerProps> = () => {
  const [state, dispatch] = useReducer(reducer, {
    games: [],
  });

  useEffect(() => {
    getOngoingGames().then((res) => {
      dispatch({ type: "GET_GAMES", payload: res });
    });

    watchGames((subscriptionData) => {
      if (subscriptionData.verb === "updated") {
        dispatch({ type: "UPDATE_GAME", payload: subscriptionData.data });
      } else if (subscriptionData.verb === "created") {
        dispatch({ type: "CREATE_GAME", payload: subscriptionData.data });
      }
    });
  }, []);

  return <GamePreviewsList games={state.games} />;
};
