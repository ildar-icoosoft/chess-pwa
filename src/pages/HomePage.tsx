import React, { FC } from "react";
import OngoingGamesContainer from "../containers/OngoingGamesContainer";
import { Button, Modal } from "react-bootstrap";
import { ChallengeAiFormContainer } from "../containers/ChallengeAiFormContainer";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../app/rootReducer";
import {
  hideChallengeAiModal,
  showChallengeAiModal,
} from "../redux/slices/challengeAiModalSlice";
import { AppDispatch } from "../app/store";

const HomePage: FC<unknown> = () => {
  const dispatch = useDispatch<AppDispatch>();

  const { isChallengeAiModalVisible } = useSelector(
    (state: RootState) => state.challengeAiModal
  );

  return (
    <>
      <Modal
        show={isChallengeAiModalVisible}
        onHide={() => dispatch(hideChallengeAiModal())}
        animation={false}
      >
        <Modal.Header closeButton>
          <Modal.Title>Play with the computer</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <ChallengeAiFormContainer />
        </Modal.Body>
      </Modal>
      <Button
        variant="primary"
        onClick={() => dispatch(showChallengeAiModal())}
      >
        Play with the computer
      </Button>
      <OngoingGamesContainer />
    </>
  );
};

export default HomePage;
