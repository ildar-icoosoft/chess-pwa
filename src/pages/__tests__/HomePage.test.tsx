import React from "react";
import TestRenderer from "react-test-renderer";
import mountTest from "../../tests/mountTest";
import HomePage from "../HomePage";
import { OngoingGamesContainer } from "../../components/OngoingGamesContainer/OngoingGamesContainer";

jest.mock("../../services/api");

describe("HomePage", () => {
  mountTest(HomePage);

  it("Snapshot", () => {
    const tree = TestRenderer.create(<HomePage />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("children components", () => {
    it("contains OngoingGamesContainer", () => {
      const testRenderer = TestRenderer.create(<HomePage />);
      const testInstance = testRenderer.root;

      expect(testInstance.findAllByType(OngoingGamesContainer).length).toBe(1);
    });
  });
});
