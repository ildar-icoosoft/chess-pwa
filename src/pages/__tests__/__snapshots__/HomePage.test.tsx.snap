// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Snapshot 1`] = `
<div
  className="grid"
/>
`;
